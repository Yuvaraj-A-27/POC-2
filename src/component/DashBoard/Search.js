import { Grid, makeStyles, TextField } from '@material-ui/core'
import React, { useState } from 'react'
import SearchIcon from '@material-ui/icons/Search';
import { connect } from 'react-redux';
import ProductCard from '../util/ProductCard';
import Carousel from 'react-elastic-carousel'

const breakPoints = [
    {width:1,itemsToShow : 1},
    {width:550,itemsToShow : 2},
    {width:768,itemsToShow : 3},
    // {width:1200,itemsToShow : 4},
]


const useStyles = makeStyles((theme)=>({
    grid :{
        padding: theme.spacing(3),
    }
}))

function Search(props){
    const classes = useStyles()

    //state
    const [search, setSearch] = useState('')

    let result = props.productDetail? props.productDetail.filter(e=>{
        const title = e.title.toLowerCase()
        const description = e.description.toLowerCase()
        return (title.includes(search.toLowerCase()) || description.includes(search.toLowerCase()))
    }) : null
    // console.log(result);
    
    const resultProduct =result? result.map(e=>(<ProductCard product_id ={e.id} title={e.title} image={e.image}/>)) : null

    let carousel = null
    if(result && result.length<15 && search.length>0 && resultProduct.length!==0){
        carousel=<Carousel breakPoints={breakPoints}>
            {resultProduct}    
        </Carousel>
    }
    

    let notFound = null
    if(result && resultProduct.length===0 && search.length >0){
        notFound = <p data-testid = 'not-found'>Oops! Sorry, Your Search Product is not Available</p>
    }
    
    

    return(
        <div>
            <Grid container spacing={1} alignItems='center' justify= 'center' className={classes.grid}>
                <Grid item >
                    <SearchIcon />
                </Grid>
                <Grid item xs='5' >
                    <TextField
                        label="Search for Product"
                        style={{ margin: 8 }}
                        fullWidth
                        size = 'medium'
                        margin="dense"
                        variant="outlined"
                        value = {search}
                        onChange = {e=>setSearch(e.target.value)}
                        // data-testid = 'search-input'
                    />
                </Grid>
            </Grid>
            {/* {result.length<15 && search.length>0 && resultProduct.length!==0 &&
                <Carousel breakPoints={breakPoints}>
                    {resultProduct}    
                </Carousel>
            } */}
            {carousel}
            {/* {resultProduct.length===0 && search.length >0 &&
                <p>Oops! Sorry, Your Search Product is not Available</p>
            } */}
            {notFound}
            
        </div>
    )
}

const mapStateToProps = state =>{
    return{
        productDetail : state.product
    }
}

export default connect(mapStateToProps,null)(Search)