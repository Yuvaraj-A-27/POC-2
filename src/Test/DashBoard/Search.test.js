import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { fireEvent, getByTestId, render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import Search from '../../component/DashBoard/Search'
import userEvent from '@testing-library/user-event'


const middleware  = {}
const mockStore = configureStore(middleware)
const initialState = {
    product : [
    {
        category: "men's clothing",
        description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
        id: 1,
        image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        price: 109.95,
        title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
    },
    {
        category: "Women's clothing",
        description: "Your men perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
        id: 2,
        image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        price: 200,
        title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
    }
    ]
}
const Store = mockStore(initialState)


describe('DashBoard Search Test', ()=>{


    test('Search snapshot test',()=>{
        const search = renderer.create(
            <Provider store={Store}>
                <Search />
            </Provider>
        ).toJSON()
        expect(search).toMatchSnapshot()
    })





    test('Search Input Functionality Test',()=>{
        const {getByPlaceholderText} = render(
            <Provider store={Store}>
                <Search />
            </Provider>
        )

        const searchInput = getByPlaceholderText('Search for Product')
        fireEvent.change(searchInput, {target:{value:'shirt'}})
        expect(searchInput.value).toBe('shirt')
    })




    test('Searcg Result Testing', ()=>{
        const {getByPlaceholderText, getByTestId} = render(
            <Provider store={Store}>
                <Search />
            </Provider>
        )
        
        const search = getByPlaceholderText('Search for Product')
        userEvent.type(search, "men")
        expect(getByTestId('productCard')).toBeInTheDocument()
    })
})