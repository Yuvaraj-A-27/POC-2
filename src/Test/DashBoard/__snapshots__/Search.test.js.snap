// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashBoard Search Test Search snapshot test 1`] = `
<div>
  <div
    className="MuiGrid-root makeStyles-grid-1 MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        style={
          Object {
            "margin": 8,
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
          data-shrink={false}
        >
          Search for Product
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
            disabled={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Search for Product"
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-4"
            >
              <span>
                Search for Product
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
